frontend/src/services/FileHandlerService.ts

```diff
@@ -0,0 +1,169 @@
+ * @param file The file to generate an md5 hash for
+ */
+const computeChecksumMd5 = (file: File): Promise<string> => {
+  return new Promise((resolve, reject) => {
+    const chunkSize = 2097152 // Read in chunks of 2MB
+    const spark = new SparkMD5.ArrayBuffer()
+    const fileReader = new FileReader()
+
+    let cursor = 0 // current cursor in file
+
+    fileReader.onerror = () => {
+      reject('MD5 computation failed - error reading the file')
+    }
+
+    // read chunk starting at `cursor` into memory
+    const processChunk = (chunk_start: number): void => {
+      const chunk_end = Math.min(file.size, chunk_start + chunkSize)
+      fileReader.readAsArrayBuffer(file.slice(chunk_start, chunk_end))
+    }
+
+    fileReader.onload = (e) => {
+      const arrayBuffer = e.target?.result
+      if (!(arrayBuffer instanceof ArrayBuffer)) {
+        return reject('Empty array buffer - error reading the file')
+      }
+      spark.append(arrayBuffer) // Accumulate chunk to md5 computation
+      cursor += chunkSize // Move past this chunk
+
+      if (cursor < file.size) {
+        // Enqueue next chunk to be accumulated
+        return processChunk(cursor)
+      } else {
+        // Computation ended, last chunk has been processed. Return as Promise value.
+        // This returns the base64 encoded md5 hash, which is what cloud services expect.
+        return resolve(btoa(spark.end(true)))
+      }
+    }
+
+    return processChunk(0)
+  })
+}
+
+const generatePresignedData = async (
+  url: PresignUrlEndpoint,
+  params: { fileId: string; fileMd5Hash: string; fileType: string },
+): Promise<PresignedData> => {
+  return ApiService.post<PresignedData>(url, params).then(({ data }) => data)
+}
+
+const postToPresignedUrl = async (
+  presignedUrl: string,
+  formData: FormData,
+): Promise<AxiosResponse<never>> => {
+  return ApiService.post(presignedUrl, formData, {
+    headers: { 'Content-Type': '' },
+    withCredentials: false,
+  })
+}
+
+/**
+ * Uploads a file by requesting a presignedUrl and posting to it.
+ *
+ * @param arg.url the url to fetch presigned data from
+ * @param arg.file the file to upload
+ * @param arg.fileId the identifier of the file
+ */
+const uploadFile = async ({
+  url,
+  file,
+  fileId,
+}: {
+  url: PresignUrlEndpoint
+  file: File
+  fileId: string
+}): Promise<UploadedFileData> => {
+  const fileMd5Hash = await computeChecksumMd5(file)
+  const presignedDataParams = {
+    fileId,
+    fileMd5Hash,
+    fileType: file.type,
+  }
+
+  const postData = await generatePresignedData(url, presignedDataParams)
+
+  // Generate formdata to post to the presigned url.
+  const formData = new FormData()
+  Object.entries(postData.fields).forEach(([k, v]) => formData.append(k, v))
+  formData.append('file', file)
+
+  // POST generated formData to presigned url.
+  const response = await postToPresignedUrl(postData.url, formData)
+
+  const encodedFileId = encodeURIComponent(fileId)
```

Comment:
Why do we encode `fileId` here? Encoding should only be done at the point of usage, where we know what encoding we actually need 🤔

Reply:
The encoded ids is just for storing the uploaded metadata in the database as the key to retrieve the data

Reply:
That sounds a bit funny, the value shouldn't be stored URI-encoded in the DB for querying 🤔 . URI encoding is a transport encoding to transmit data in a URL-safe fashion.

Well, it's probably too late to change this for consistency if there are encoded entries in the DB already, but we can keep it in mind for next time 😃